client/network/src/request_responses.rs

```diff
@@ -0,0 +1,866 @@
+					match outcome {
+						RequestProcessingOutcome::PendingResponse {
+							protocol, inner_channel, response
+						} => {
+							if let Some((protocol, _)) = self.protocols.get_mut(&*protocol) {
+								protocol.send_response(inner_channel, response);
+							}
+						}
+						RequestProcessingOutcome::Busy { peer, protocol } => {
+							let out = Event::InboundRequest {
+								peer,
+								protocol,
+								outcome: Err(InboundError::Busy),
+							};
+							return Poll::Ready(NetworkBehaviourAction::GenerateEvent(out));
+						}
+					}
+				}
+			}
+
+			// Poll request-responses protocols.
+			for (protocol, (behaviour, resp_builder)) in &mut self.protocols {
+				while let Poll::Ready(ev) = behaviour.poll(cx, params) {
+					let ev = match ev {
+						// Main events we are interested in.
+						NetworkBehaviourAction::GenerateEvent(ev) => ev,
+
+						// Other events generated by the underlying behaviour are transparently
+						// passed through.
+						NetworkBehaviourAction::DialAddress { address } => {
+							return Poll::Ready(NetworkBehaviourAction::DialAddress { address })
+						}
+						NetworkBehaviourAction::DialPeer { peer_id, condition } => {
+							return Poll::Ready(NetworkBehaviourAction::DialPeer {
+								peer_id,
+								condition,
+							})
+						}
+						NetworkBehaviourAction::NotifyHandler {
+							peer_id,
+							handler,
+							event,
+						} => {
+							return Poll::Ready(NetworkBehaviourAction::NotifyHandler {
+								peer_id,
+								handler,
+								event: ((*protocol).to_string(), event),
+							})
+						}
+						NetworkBehaviourAction::ReportObservedAddr { address } => {
+							return Poll::Ready(NetworkBehaviourAction::ReportObservedAddr {
+								address,
+							})
+						}
+					};
+
+					match ev {
+						// Received a request from a remote.
+						RequestResponseEvent::Message {
+							peer,
+							message: RequestResponseMessage::Request { request, channel },
+						} => {
+							let (tx, rx) = oneshot::channel();
+
+							// Submit the request to the "response builder" passed by the user at
+							// initialization.
+							if let Some(resp_builder) = resp_builder {
+								// If the response builder is too busy, silently drop `tx`.
+								// This will be reported as a `Busy` error.
+								let _ = resp_builder.try_send(IncomingRequest {
+									origin: peer.clone(),
+									request_bytes: request,
+									answer: tx,
+								});
+							}
+
+							let protocol = protocol.clone();
+							self.pending_responses.push(Box::pin(async move {
+								// The `tx` created above can be dropped if we are not capable of
+								// processing this request, which is reflected as a "Busy" error.
+								if let Ok(response) = rx.await {
+									RequestProcessingOutcome::PendingResponse {
+										protocol, inner_channel: channel, response
+									}
+								} else {
+									RequestProcessingOutcome::Busy { peer, protocol }
```

Comment:
I didn't realize that the unexpected EOF would lead to the connection being closed.

Being able to write an error response is something I've considered.
However we would ideally change the current request-response-like protocols (block requests, finality proof requests, light client requests) to use this new request-response behaviour as their underlying implementation, which wouldn't be possible if we add extra bytes around the response.

If that sounds good, I would instead change this PR to treat an EOF before receiving the length prefix as a non-fatal error.

Reply:
> If that sounds good, I would instead change this PR to treat an EOF before receiving the length prefix as a non-fatal error.

You mean in `RequestResponseCodec::read_response`? Sounds good to me, then `type Response = Option<Vec<u8>>` or similar, right?

Reply:
> I didn't realize that the unexpected EOF would lead to the connection being closed.

Essentially any I/O error that occurs in an inbound or outbound upgrade is considered to be fatal as far as `libp2p-request-response` goes, because it is not really possible to differentiate - it is up to the `RequestResponseCodec` implementation. Only upgrade timeouts are non-fatal for the connection.

Reply:
> You mean in RequestResponseCodec::read_response? Sounds good to me, then type Response = Option<Vec<u8>> or similar, right?

Yes, I had `type Response = Result<Vec<u8>, ...>;` in mind